### ISSUE TITLE
GList: signature refers to itself

### ISSUE THREAD

**Issue by ManasJayanth** [cedlemo/OCaml-GLib2#3]
I tried to generate bindings for `GList` and got the following

```ml
open Ctypes
open Foreign

type t
let t_typ : t structure typ = structure "List"

let f_data = field t_typ "data" (ptr void)
let f_next = field t_typ "next" (ptr List.t_typ)
let f_prev = field t_typ "prev" (ptr List.t_typ)
let _ = seal t_typ
```

Looks like that should have be `t_typ`. Dune complains `The module List is an alias for module GLib__List, which is missing`

**Comment by cedlemo** [comment]
@prometheansacrifice ,

sorry for the delay, I guess that you tried to add GList in the gen.ml file., the GList bindings already exists in https://github.com/cedlemo/OCaml-GLib2/blob/master/lib/DLList.ml and the SLit bindings in https://github.com/cedlemo/OCaml-GLib2/blob/master/lib/SLList.ml. I changed the name GList to DLList (doubly-linked-list) and SLList (Singly-linked-list) to avoid name clash.

You remember the mail I send you in which I said that generating recursive Ctypes structures are not trivial, so in order to keep my bindings generator simple and adaptable, I decided that it should works like this:
* define the data/structures, elements that I want to automaticaly generate the bindings in the gen.ml file
* add manualy elements that I want to "tweak" the bindings in the lib repository.

For now, I just have done the bindings for:
* GList
* SList
* GHashTable

The bindings are in `lib` and you should be able to see how to use them in the tests.

Of course, I will be happy to hear your thoughts about this or the ideas you can have to improve my bindings generator.

**Comment by ManasJayanth** [comment]
I haven't looked deeper, but can the following be generated instead?

```ml
open Ctypes
open Foreign

type t
let t_typ : t structure typ = structure "List"

let f_data = field t_typ "data" (ptr void)
let f_next = field t_typ "next" (ptr t_typ)
let f_prev = field t_typ "prev" (ptr t_typ)
let _ = seal t_typ
```
If you think its feasible, I'd dig deeper and raise a PR

**Comment by ManasJayanth** [comment]
When a datastructure is is being queried and if one of the returned methods have params/return types same as the datastructure itself, it should be possible to use `t` instead of say `List.t`

This is different from the cyclic deps issue I had mentioned earlier. This seem self referential.

**Comment by cedlemo** [comment]
> When a datastructure is is being queried and if one of the returned methods have params/return types same as the datastructure itself, it should be possible to use t instead of say List.t

Yes you are totally right, but I would prefer that you come with an example from another lib than GLib.

The case of GLib is really specific because GLib defines all the types / containing types(List, Arrray, HashTable) for all the libs that can be "gobject-instrospected". That is why I don't really want to modify gi-ctypes-bindings-generator based on some needs for GLib. I prefer to keep it simple and do a custom gen.ml for GLib. (I hope I am clear enough, it is not easy to explain).

For example, did you find an example of a recursive data structure in another lib like in https://github.com/cedlemo/ges-bindings-ocaml/blob/master/gen/gen.ml for the `GES` namespace? It would be better to use it to test and modify the gi-ctypes-generator. The fact is GLib is the less generic libs of all the libs that can be introspected.

Furthermore, if you look further in https://github.com/cedlemo/OCaml-GLib2/blob/master/lib/DLList.ml, beside the simple structure definition, there are some code related to methods with callback which will be hard to reproduce with the binding generator.

**Comment by ManasJayanth** [comment]
I see. Well, when I generated bindings for GES, some of the methods depend on GList. This is what motivated me to generate it myself. I guess I'll have to eject them out.

At the moment, I don't have another instance of this happening, so I'll close this for now. I'll re-open if I do.

**Comment by ManasJayanth** [comment]
@cedlemo This means the bindings generator will have to handle methods depending on `GList` differently - it will have to appropriately assign singly/doubly linked list. 

I think it will be easier to just call it `List` and let consumers `module GList = GLib.List`

**Comment by cedlemo** [comment]
Please, feel free to try it.
